mypy/literals.py

```diff
@@ -113,7 +113,10 @@ def visit_star_expr(self, e: StarExpr) -> Key:
         return ('Star', literal_hash(e.expr))
 
     def visit_name_expr(self, e: NameExpr) -> Key:
-        return ('Var', e.name)
+        # N.B: We use the node itself as the key, and not the name,
+        # because using the name causes issues when there is shadowing
+        # (for example, in list comprehensions).
+        return ('Var', e.node)
```

Comment:
Shouldn't we instead push a new binber frame when accepting comprehension to match what Python does?

I mean binder works fine with:
```python
def foo(x: object) -> None:
    if isinstance(x, str):
        def nested() -> None:
            x = 42
            reveal_type(x)
```

Reply:
That would discard other bindings, right? Which we wouldn't want.

Reply:
Indeed, this may cause tons of new false positives (unlike for normal functions where this is questionable, comprehensions are functions executed immediately after definition so always safe).

OK, although this somehow looks dangerous on some unconscious level, I can't really object, so go ahead.

Reply:
Yeah, it is a little perilous seeming, but the `Var` node is fundamentally how we represent a variable binding, and having separate `Var` nodes is how we implement the new scope in a comprehension, so I think it is OK.